src/TestSuite/IntegrationTestTrait.php

```diff
@@ -669,11 +669,25 @@ protected function _addTokens(string $url, array $data): array
 
         if ($this->_csrfToken === true) {
             $middleware = new CsrfProtectionMiddleware();
-            if (!isset($this->_cookie['csrfToken'])) {
-                $this->_cookie['csrfToken'] = $middleware->createToken();
+            $token = null;
+            if (!isset($this->_cookie['csrfToken']) && !isset($this->_session['csrfToken'])) {
+                $token = $middleware->createToken();
+            } elseif (isset($this->_cookie['csrfToken'])) {
+                $token = $this->_cookie['csrfToken'];
+            } else {
+                $token = $this->_session['csrfToken'];
+            }
+
+            // Add the token to both the session and cookie to cover
+            // both types of CSRF tokens. We generate the token with the cookie
+            // middleware as cookie tokens will be accepted by session csrf, but not
+            // the inverse.
+            if ($token) {
```

Comment:
In what scenario is this not set here?

Reply:
You're right it will always be set, I was being overly cautious.